### ISSUE TITLE
SNS/SQS failing under load (akka.actor.DeadLetterActorRef)

### ISSUE THREAD

**Issue by kbkk** [localstack/localstack#2074]
Hello,

I'm running latest localstack (with #1986 patch), and I have ~20 services listening to a total of ~200 SQS queues (~20 each) and producing to ~40 SNS topics. 

Basically, under load (30 events a second) the console is full of:

```
localstack_1                | 2020-02-20T15:19:49:WARNING:urllib3.connectionpool: Connection pool is full, discarding connection: 4bc7a10e1c4b
```

When I enable DEBUG=1 I also get some:

```
localstack_1                | 14:42:19.408 [elasticmq-akka.actor.default-dispatcher-30] INFO  akka.actor.DeadLetterActorRef - Message [org.elasticmq.MessageData] from Actor[akka://elasticmq/user/$a/$h#880680888] to Actor[akka://elasticmq/deadLetters] was not delivered. [10] dead letters encountered, no more dead letters will be logged. If this is not an expected behavior, then [Actor[akka://elasticmq/deadLetters]] may have terminated unexpectedly, This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```

Most of messages are undelivered from SNS to SQS.

What I tried (did not help):

* replace `maxsize=128` up to `maxsize=1024` https://github.com/localstack/localstack/blob/0a73a20b7706ad999c961ed012fd8306891041ed/localstack/services/infra.py#L424
* add `akka.http.server.request-timeout = 40 s` option https://github.com/localstack/localstack/blob/793be6c298ca9a65f162f1f9c9d38ffab2dbe45f/localstack/services/sqs/sqs_starter.py#L39

Also I don't have any simple way to reproduce the problem :(

Edit:
Applying this patch https://github.com/boto/botocore/issues/619#issuecomment-583511406
seems to work for both mentioned errors but messages are still lost.

Edit #2:
I think my main problem is that localstack's SNS subscription are sent synchronously. My app works way better on s12v/sns.

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-155) by [Unito](https://www.unito.io/learn-more)

**Comment by Zambonilli** [comment]
We're seeing localstack hang on SQS calls when trying to write 500k < 1KiB messages to SQS. We are using `localstack/localstack-light:0.11.1` image and are throttling the calls in the client. At ~420k messages queued, the localstack proxy can't connect to the sqs anymore.

We were hitting a limit to the elasticmq java heap limit that looks to be hardcoded to 256MiB. I updated this to 3g in the sqs_starter.py file in the docker container. From there, I kill -9 all the processes localstack start launches in the container and then finally kill -9 the localstack process and it looks like make infra or supervisord restarts the localstack process.

https://github.com/localstack/localstack/blob/v0.11.1/localstack/services/sqs/sqs_starter.py#L13

**Comment by whummer** [comment]
Thanks for reporting @kbkk @Zambonilli . We have recently refactored the API proxy layer with a new HTTP server implementation (HTTP/2 capable, among other things). Can you please give it a try with the `latest` version of the Docker image?

We can definitely make the hardcoded heap space limit of 256MiB configurable. Regarding the SNS->SQS integration - would be great if you could provide a small, isolated example to reproduce the issue. Thanks for your help.

**Comment by kbkk** [comment]
@whummer Thanks for taking interest in the issue.

I've decided to try out the `latest` (and `0.11.3`) image with out integration test suite, and the performance is WAY worse.

Creating 20 queues in parallel (test setup):
- `0.10.7`: 300ms
- `0.11.3`: 3700ms

Purging and deleting them in parallel (test teardown):
- `0.10.7`: 750ms
- `0.11.3`: 3050ms

**Comment by AdallomRoy** [comment]
Experiencing this as well, SNS implementation is definitely underwhelming...

**Comment by kgston** [comment]
You might want to revert back to the old SQS backend (elasticmq) using the env var `SQS_PROVIDER=elasticmq`. We faced the same performance issues and reverting back helped a lot.

**Comment by kbkk** [comment]
It's definitely SNS / SQS integration that was holding me back. IIRC what I investigated back when I created this issue was the python sqs client was synchronous - hence the slowdown

**Comment by nhoughto** [comment]
So my workaround for this is to use goaws for SNS and SQS (https://github.com/p4tin/goaws), its in golang and pretty performant. ElasticMQ is fine but it only does SQS. The approach is basically to separate the SQS and SNS topics out of localstack, so localstack doesn't even know about them directly, so you have to ensure SQS/SNS API calls go to `goaws` rather than localstack, then configure localstack with `TEST_SQS_URL` so that localstack lambda events can still flow to goaws. Its slightly less obvious than SQS_BACKEND=elasticmq because the API calls aren't proxied, but waaay more performant.. because the API calls aren't proxied..

**Comment by thrau** [comment]
Hello everyone, I'm closing this issue since most information is outdated. The performance issues with sqs should be solved by now. Networking issues could be related to #5499 but I can't say without a reproducible example. Feel free to open a new issue if the problem persists.
